pkg/expr/graph.go

```diff
@@ -211,19 +211,29 @@ func (s *Service) buildDependencyGraph(req *Request) (*simple.DirectedGraph, err
 
 // buildExecutionOrder returns a sequence of nodes ordered by dependency.
 // Note: During execution, Datasource query nodes for the same datasource will
-// be grouped into one request and executed first as phase after this call.
+// be grouped into one request and executed first as phase after this call
+// If the groupByDSFlag is enabled.
 func buildExecutionOrder(graph *simple.DirectedGraph) ([]Node, error) {
 	sortedNodes, err := topo.SortStabilized(graph, nil)
 	if err != nil {
 		return nil, err
 	}
 
-	nodes := make([]Node, len(sortedNodes))
-	for i, v := range sortedNodes {
-		nodes[i] = v.(Node)
+	var dsNodes []Node
+	var otherNodes []Node
+
+	for _, v := range sortedNodes {
+		n := v.(Node)
+		switch n.NodeType() {
+		case TypeDatasourceNode:
```

Comment:
ML node should be together with ds nodes as it's effectively a ds node but executed by ML